cpp/src/arrow/compute/exec/options.h

```diff
@@ -294,6 +294,20 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
       this->key_cmp[i] = JoinKeyCmp::EQ;
     }
   }
+  HashJoinNodeOptions(
```

Comment:
Rather than keep adding constructors for `HashJoinNodeOptions` can we make a constructor that takes no arguments (or the minimum number of required arguments?  Maybe just `left_keys` and `right_keys` and assigns reasonable defaults).  Then we can just override the settings after construction.  For example:

```
HashJoinNodeOptions opts{left_keys, right_keys};
opts.join_type = join_type_from_substrait;
...
```